internal/controllers/module_nmc_reconciler.go

```diff
@@ -81,31 +86,26 @@ func (mnr *ModuleNMCReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		return ctrl.Result{}, fmt.Errorf("failed to set finalizer on %s Module: %v", req.NamespacedName, err)
 	}
 
-	// get all nodes
-	nodes, err := mnr.reconHelper.getNodesList(ctx)
+	// get nodes targeted by selector
+	targetedNodes, err := mnr.reconHelper.getNodesListBySelector(ctx, mod)
 	if err != nil {
-		return ctrl.Result{}, fmt.Errorf("failed to get nodes list: %v", err)
+		return ctrl.Result{}, fmt.Errorf("failed to get list of nodes by selector: %v", err)
 	}
 
-	errs := make([]error, 0, len(nodes))
-	for _, node := range nodes {
-		kernelVersion := strings.TrimSuffix(node.Status.NodeInfo.KernelVersion, "+")
-		mld, err := mnr.kernelAPI.GetModuleLoaderDataForKernel(mod, kernelVersion)
-		if err != nil && !errors.Is(err, module.ErrNoMatchingKernelMapping) {
-			logger.Info(utils.WarnString(fmt.Sprintf("internal errors while fetching kernel mapping for version %s: %v", kernelVersion, err)))
-			errs = append(errs, err)
-			continue
-		}
-		shouldBeOnNode, err := mnr.reconHelper.shouldModuleRunOnNode(node, mld)
-		if err != nil {
-			logger.Info(utils.WarnString(fmt.Sprintf("failed to determine if module %s/%s should be on node %s: %v", mld.Namespace, mld.Name, node.Name, err)))
-			errs = append(errs, err)
-			continue
-		}
-		if shouldBeOnNode {
-			err = mnr.reconHelper.enableModuleOnNode(ctx, mld, &node, kernelVersion)
+	currentNMCs, err := mnr.reconHelper.getNMCsByModuleSet(ctx, mod)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get NMCs for %s Module: %v", req.NamespacedName, err)
+	}
+
+	sdMap, err := mnr.reconHelper.prepareSchedulingData(ctx, mod, targetedNodes, currentNMCs)
+	errs := make([]error, 0, len(sdMap)+1)
+	errs = append(errs, err)
+
+	for nodeName, sd := range sdMap {
```

Comment:
If `prepareSchedulingData` returns an error, we should exit the function and not use `sdMap`.

Reply:
Why? We are going over multiple nodes here. The error may be in one of the nodes, but for the rest the sdMap entries were built successfully. So, we will handle the entries, and will return an error, and the reconcile loop will run again